src/main/java/com/thebois/models/beings/Colony.java

```diff
@@ -1,25 +1,133 @@
+        if (idleBeing.isPresent()) {
+            final AbstractRole role = RoleFactory.fromType(roleType);
+            idleBeing.get().setRole(role);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public boolean tryIncreaseAllocation(final RoleType roleType, final int amount) {
+        if (!canIncreaseAllocation(amount)) return false;
+        for (int i = 0; i < amount; i++) {
+            if (!tryIncreaseAllocation(roleType)) return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean tryDecreaseAllocation(final RoleType roleType) {
+        final AbstractRole role = RoleFactory.fromType(roleType);
+        final Optional<IBeing> beingWithRole = findBeingsWithRole(role).stream().findFirst();
+        if (beingWithRole.isPresent()) {
+            beingWithRole.get().setRole(null);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean tryDecreaseAllocation(final RoleType roleType, final int amount) {
+        if (!canDecreaseAllocation(roleType, amount)) return false;
+        for (int i = 0; i < amount; i++) {
+            if (!tryDecreaseAllocation(roleType)) return false;
+        }
+        return true;
+    }
+
+    @Override
+    public int countBeingsWithRole(final RoleType roleType) {
+        final AbstractRole role = RoleFactory.fromType(roleType);
+        return (int) getBeingsByReference().stream()
+                                           .filter(being -> role.equals(being.getRole()))
+                                           .count();
+    }
+
+    @Override
+    public boolean canDecreaseAllocation(final RoleType roleType) {
+        return canDecreaseAllocation(roleType, 1);
+    }
+
+    @Override
+    public boolean canDecreaseAllocation(final RoleType roleType, final int amount) {
+        final AbstractRole role = RoleFactory.fromType(roleType);
+        return findBeingsWithRole(role).size() >= amount;
+    }
+
+    @Override
+    public boolean canIncreaseAllocation() {
+        return canIncreaseAllocation(1);
+    }
+
+    @Override
+    public boolean canIncreaseAllocation(final int amount) {
+        return findIdleBeings().size() >= amount;
+    }
+
+    private Collection<IBeing> findIdleBeings() {
+        final Stream<IBeing> idleBeings = getBeingsByReference().stream().filter(this::isIdle);
+        return idleBeings.collect(Collectors.toList());
+    }
+
+    private boolean isIdle(final IBeing being) {
+        return being.getRole() == null;
+    }
+
+    private Collection<IBeing> findBeingsWithRole(final AbstractRole role) {
+        return getBeingsByReference().stream()
+                                     .filter(being -> being.getRole() != null)
+                                     .filter(being -> role.equals(being.getRole()))
+                                     .collect(Collectors.toList());
```

Comment:
This could be ColonyRoleManager. Right now its borderline OK imo but if we need more in the future we should move these

Reply:
That's what I thought as well. My original implementation had `Colony` use a `RoleAllocator`. I encountered problems though, because the `RoleAllocator` only operates directly on the `IBeing` instances inside `Colony`. That's a code smell, which points towards that it is `Colony`s responsibility to handle the role management.

I actually asked about this in our Slack and you guys said to just put it in `Colony` 😛

Reply:
For now its OK, if we add more stuff and Colony actually does something we will need to look into a solution.